Small bug fix to the StereoEnhancer effect and Gain knob

IanCaio opened the pull request:
Labeling it as `after-refactor`, though it has a small bug fix and should be a really quick one to review. Only affects the StereoEnhancer plugin.

- Fixes a small bug where the buffer was never cleared because the conditional to clearing it was never reached.

- Improves the clearing conditional so it only clears the buffer if it was actually affected in the previous process cycle.

- Adds a Gain knob to be able to adjust the level of the output signal so a more appropriate comparison of the before/after sound is done.

IanCaio committed:
Improvements to the StereoEnhancer effect

	Fixes a small bug where the buffer was never cleared because the
conditional to clearing it was never reached.
	Adds a Gain knob to be able to adjust the level of the output
signal so a more appropriate comparison of the before/after sound is
done.

Spekular commented:
Since `clear` is both a verb and an adjective, the variable name is a bit confusing. Can we think of something unambiguous instead?
```
bufferClear, bufferCleared, bufferIsClear
shouldClearBuffer, mustClearBuffer
filledBuffer, fillBuffer
```

Veratil commented:
> 
> 
> Since `clear` is both a verb and an adjective, the variable name is a bit confusing. Can we think of something unambiguous instead?
> 
> ```
> bufferClear, bufferCleared, bufferIsClear
> shouldClearBuffer, mustClearBuffer
> filledBuffer, fillBuffer
> ```

Would `empty` be clearer? (no pun intended lol)

IanCaio commented:
Sure, now that I read it again name is a bit ambiguous. Everyone fine with `m_emptyBuffer`?

Spekular commented:
I skipped that one intentionally because empty is another example of a verb and adjective, so it's no different from clear.

IanCaio committed:
Changes variable name and fix its use

	Changes the variable name that tells whether the buffer is clear
and actually sets it to true on the clearBuffer method.

IanCaio commented:
On second thought, personally I think since the prefix and lack of parenthesis makes clear it's a variable it isn't that confusing. But I went and changed it to `m_isBufferClear`, which looks good too IMO. I also forgot to set it to true on the clear method, it's fixed now.

Veratil commented:
> 
> 
> On second thought, personally I think since the prefix and lack of parenthesis makes clear it's a variable it isn't that confusing. But I went and changed it to `m_isBufferClear`, which looks good too IMO. I also forgot to set it to true on the clear method, it's fixed now.

I think `bufferIsClear` reads better than `isBufferClear`. `if (m_isBufferClear)` reads as `if is buffer clear` while `if (m_bufferIsClear)` reads as `if buffer is clear`.

Spekular commented:
> the prefix and lack of parenthesis makes clear it's a variable

I didn't mean that it would be confused for a method, I meant that the purpose of the variable could be misunderstood (inverted)
- clearBuffer = true, as in "the buffer is clear"
- clearBuffer = true, as in "we should/will clear the buffer (because it isn't clear right now)"

+1 for Veratil's point about the naming in if-statements.

IanCaio committed:
Changes variable name to a better one

IanCaio commented:
> I didn't mean that it would be confused for a method, I meant that the purpose of the variable could be misunderstood (inverted)
> 
>     * clearBuffer = true, as in "the buffer is clear"
> 
>     * clearBuffer = true, as in "we should/will clear the buffer (because it isn't clear right now)"

Oh, I see it now! You're right, it can be very misleading.

I changed it again to Veratil's suggestion!

Veratil commented:
To the same point of readability, is buffer `clear` or is buffer `empty` a better read?

IanCaio commented:
I think either is fine, but `clear` has the advantage of being consistent with the method name.

PhysSong commented:
Why you didn't implement loading/saving `m_outputGain`?

IanCaio commented:
> Why you didn't implement loading/saving `m_outputGain`?

Oh, I think I forgot it :sweat_smile: 

I'll add it and push it later today

IanCaio committed:
Includes the output gain on save/load

	Saves and loads the new output gain control on the project file
now.

IanCaio commented:
Now outputGain is loaded and saved on the project file!

zonkmachine commented:
Looks like it's pretty much done from judging by the comments. A quick rebase/merge?
@IanCaio Are you still around? :whale:

PhysSong committed:
Merge branch 'master' into feature/stereoEnhancerImprov

zonkmachine committed:
Merge branch 'master' into feature/stereoEnhancerImprov

zonkmachine committed:
KnobType::Bright26

zonkmachine commented:
I resolved a merge conflict via the web-editor. It seem to have failed to merge correctly anyway and the last manual change should fix that.

messmerd commented:
The buffer clear bug was fixed in #8070, so this PR could focus on just the gain knob now.

JohannesLorenz commented:
As @IanCaio left LMMS, I propose to take over the PR and care about the knob (obviously this is "after feature freeze", but can yet be prepared). Everyone OK with this?
